Week of Jan 6 changes (#222)
xregistry/spec: core/spec.md

a way that it can be represented in a single file, including but not limited
to JSON, or via the entity graph of a REST API.

If an error occurs during the processing of a request, even if the error was
during the creation of the response (e.g. an invalid `?inline` value was
provided), then an error MUST be generated and the entire request (even any

> update actions) MUST be undone. In general, when an error is generated it

Reviewer: Is there a transactionality element here, in terms of requiring that no changes applied due to the processing of the request must be observable until the processing has completed? (i.e. you can't just start making changes which could be observed, then revert them later).

Author: I kind of think there has to be, at least on a per API call basis, otherwise we run the risk of people seeing a Registry that's in a bad/invalid state.

Perhaps we could add a section to the spec (or primer) to talk about how:
- our expectation is that there's a transaction system in place where each individual API call is processed within the scope of one Tx such that if something goes wrong during ANY of its processing then entire thing can be rolled back
- this also then implies that any changes being made within that one Tx are not visible to other API calls (Txs) until it is committed 
- if someone wants to expose parts of their "desired set of changes" in an incremental fashion, they can do so by breaking up the work into individual API calls. 

Thoughts?

> SHOULD be sent back to the client. However, this MAY not happen if the server

Reviewer: Presumably if an error occurs, the server still has to send *an* error back to the client, but it doesn't need to have full information?

Author: I think we need to leave it as an impl choice since there are some systems that might think it's best to just close the connection rather than even give any error message at all - again for security reasons. I think this could be one of those "self correcting issues" where if people make poor choices then people will stop using that system. This is why I prefer to use the phrase "generate an error" rather than "send (or respond with) an error"... keeps the spec flexible.

> determines there is a good reason to not do so - such due to security concerns.

Reviewer: ```suggestion
determines there is a good reason to not do so - such as due to security concerns.
```

Author: fixed

  - See [Serializing Resource Documents](#serializing-resource-documents)
    for more information about serializing maps as HTTP headers.
- `object` - a nested entity made up of a set of attributes of these data types.

> - `relation` - a relative URL reference to a type of entity in the Registry.

Reviewer: Drive-by comment, without either knowing the motivation behind the change or having expertise in the difference between URI and URL - if we make this URI after all, we should potentially revert some of the other URI=>URL changes too.

> - `relation` - a relative URL reference to a type of entity in the Registry.

Author: the general pattern, I believe, we follow is that URLs are for HTTP GET-able addresses, while URIs are for addresses that could be concrete (like URLs) but could also be abstract and not HTTP GET-able. I believe @deissnerk wants the option for "relations" to be abstract in the future. Right Klaus?

@jskeet do you think some of our URLs need to be URIs?

> - `relation` - a relative URL reference to a type of entity in the Registry.

Reviewer: I have no opinion on what should be which - I just noted that there's a change earlier on.

I would also note that URL doesn't imply HTTP - if this is the expectation for xregistry, it may be worth documenting. (That might already be documented of course.)

> - `relation` - a relative URL reference to a type of entity in the Registry.

Author: I think we need to make this a bit clearer... we don't want to mandate http, but it is the only protocol we support today and we've been a bit lazy about using generic wording vs falling into the "we're using http" world.

I added this issue: https://github.com/xregistry/spec/issues/230 to think more about this.

  all other situations `$structure` MUST NOT be included.

- Constraints:
  - MUST be unique across all entities in the Registry.
  - MUST be a non-empty absolute URL.
  - MUST be a read-only attribute in API view.
- Examples:
  - `https://example.com/registry/endpoints/ep1`

##### `shortself` Attribute

- Type: URL
- Description: A unique absolute URL for an entity. This attribute MUST be
  an alternative URL for the owning entity's `self` attribute. The intention
  is that `shortself` SHOULD be shorter in length than `self` such that it
  MAY be used when the length of the URL referencing the owning entity is
  important. For example, in cases where the size of a message needs to be
  as small as possible.

  This specification makes no statement as to how this URL is constructed,

>   to which host/path it references, or whether an HTTP request to this URL

Reviewer: Possibly remove HTTP from here, to allow for other schemes which may also have redirection?

Author: done

  will directly perform the desired operation or whether it returns an HTTP
  redirect to the full `self` URL requiring the client to resend the request.

  If an entity is deleted and then a new entity is created that results in
  the same `self` URL, this specification does not mandate that the same
  `shorturl` be generated, but it MAY do so.

>   This attribute MUST only appear in the serialization if the `shortself`

Reviewer: This feels like a potential problem to me - or I may just have misunderstood. Consider the following sequence of events:

- Initial state: a registry with the `shortself` capability enabled
- t0: resource X is created with a long full resource URL, and a shortened URL created using URL shortener 1
- t1: registry has the `shortself` capability disabled (so the first short URL is no longer visible/updatable)
- t2: the provider of URL shortener 1 warns of turndown
- t3: the creator of the entity migrates to URL shortener 2, updating entities in all registies where the `shortself` capability is enabled - but can't update entity X
- t4: URL shortener 1 is turned down
- t5: the registry containing entity X has `shortself` re-enabled

At that point entity X is effectively broken, and the creator couldn't have pre-emptively fixed it.

Maybe this is too obscure an issue to be worth considering.

Reviewer: (When writing this comment, I hadn't got to the bit saying it has to be read-only...)

Author: Yup, this is a server controlled value so any updates to it would need to be done by the server, not the owner/creator of the entity. So the scenario as you described it is a bit wonky... however, you do raise a good point about switching to a new URL shortener service. This text was added because changing the shortself, in general, is dangerous and so I wanted it to be as static as "self", but I just realized that we don't even say that "self" is static...and it can't be for a variety of reasons. I was just trying to avoid people changing "shortself" randomly.

So, how about this:
- I'll remove the "- MUST NOT change during the lifetime of the entity." and change "static" to "known" in this paragraph... just to loosen things
- I'll add text to the Primer about how both "self" and "shorturl" are meant to be static for the lifetime of the owning entity, however there may be situations where they need to change... moving hosts, changing short-url service... so while the spec doesn't ban changing them, care should be taken when doing so since any reference using the old values might break people

I've gone ahead and made these changes since I think it's at least better than what was there before your review.

thoughts?

Reviewer: Sounds good - I'd misinterpreted who provided the shorturl; if it's server-provided, that's much simpler.

  capability is enabled. However, if this capability is enabled, then disabled,
  and then re-enabled, the `shortself` values MUST retain their original
  values. In this sense, implementations might create a `shortself` that is
  static for the lifetime of the entity and the capability controls whether
  the attribute is serialized or not.

- Constraints:
  - MUST be unique across all entities in the Registry.
  - MUST be a non-empty absolute URL referencing the same entity as the `self`
    URL, either directly or indirectly via an HTTP redirect.
  - MUST be a read-only attribute in API view.

>   - MUST NOT change during the lifetime of the entity.

Reviewer: I suspect this could cause problems - migrating from one shortening service to another feels like a reasonable use case... although I can see how anything using this effectively as an ID could be confused by changes.

Worth discussing in the subgroup with folks more knowledgeable than myself?

Author: See previous comment.... good catch!

however it is STRONGLY RECOMMENDED that server-side implementations support at
least the "read" (e.g. HTTP `GET`) operations. Implementations MAY choose to
incorporate authentication and/or authorization mechanisms for the APIs.

If an OPTIONAL HTTP path is not supported by an implementation, then any
use of that API MUST generate a `404 Not Found` error.

If a HTTP method is not supported for a defined HTTP path then a
`405 Method Not Allowed` error MUST be generated.

> Implementation MAY defined extension APIs however the following rules MUST

Reviewer: ```suggestion
Implementations MAY support defined extension APIs however the following rules MUST
```

(Or something similar...)

Author: done

apply:
- New HTTP paths that extend non-root paths MUST NOT be defined.
- New root HTTP paths MAY be defined as long as they do not use Registry level
  HTTP paths or attribute names. This include extension and GROUPS collection
  attribute names.
- Additional HTTP methods for specification defined HTTP paths MUST NOT be
  defined.

> For example, a new API with an HTTP path of `/my-api` is allowed, but APIs with

Reviewer: Doesn't this mean that the specification can never add any new root paths without that being a breaking change?
I wonder whether it's worth reserving a URI space - maybe even `/extensions/` and saying everything has to be under that?

Author: Why would adding a new root path to the spec be a breaking change, as long as it's optional ?

Or instead of "breaking" did you mean "potentially conflicting with an extension" ? Yes that can happen, but that's true of all extensions (even attribute extensions). It's just the nature of how extensions work... unless we create "extension buckets/spaces" as you suggest. But then we're back to this (very long winded) "X-" prefix on http header extensions rathole ;-)

Reviewer: > Or instead of "breaking" did you mean "potentially conflicting with an extension" ?

Yes, as I view that as a breaking change - it means that doing something that *was* valid is no longer valid. Surely that's reasonable to count as "breaking".

And yes, this is like the CloudEvents discussion we had - basically I personally thing CE went the wrong route here, as I think we've now come to the conclusion we can't add *any* core attributes without taking a 2.0. I'd like xregistry to avoid getting into that situation. (I seem to remember writing a lot about this before - https://docs.google.com/document/d/1n91IEyB8gD8JSBiAYyYo8YbDne4KD3YvCa5BrdemZt4/edit?tab=t.0#heading=h.yflgqnc5ygl0 may be what I'm thinking of, or there could be something else.)

```

The following describes the attributes of Registry model:
- `labels`
  - A set of name/value pairs that allows for additional metadata about the
    Registry to be stored without changing the schema of the model.
  - If present, MUST be a map of zero or more name/value string pairs.
  - Keys MUST be non-empty strings.
  - Values MAY be empty strings.
  - The following labels are defined for convenience and MAY also be used
    as labels for Groups or Resource model types:

>     - Name: `compatiblewith`
>       Value: A URI reference to Registry that this Registry's model is
>       compatible with. This specification does not mandate any runtime
>       validation of this claim.

Reviewer 2: Here we need to clarify how to point to a particular model, it's not sufficient to point to an xregistry itself as it could contain all kinds of models. Compatibility will always be between particular models, not between the registries as a whole.

```suggestion
    - Name: `compatiblewith`
      Value: A URI reference to an xRegistry model that this model is
      compatible with. All models that share the same reference are interoperable. 
      This specification does not mandate any runtime validation of this claim.
```

But we need to clarify probably a few more points:
* How to construct this URL? Ideally, we would point to the model, e.g. on group side we'd point from the group model to e.g. https://xregistry.cncf.org/model#groups/schemagroups (do we have a link where I can retrieve just a partial model? Otherwise we need anchor tags here. 
* What does compatible mean? 
  * It's primarily about ecosystem compatibility around a model (like we have with OpenAPI, JSON Schema of a "major" version). 
  * There's import and export compatibility, which depends on how the own model was extended:
    * Adding optional attributes: Full import and export compatibility
    * Removing optional attributes: Full import and export compatibility
    * Adding further constraints to existing attributes: Full import and export compatibility
    * Adding new mandatory attribute or making optional attribute mandatory: Full export compatibility, not import.

This is probably too much for the place here? Where would you put content that explains a concept in more detail, to not overload the spec itself?

One open question: Should we also point out a version of the model here? Most specs at least have a <major>.<minor> version, although the compatibility should be sufficiently expressed with just <major>. Maybe a second label? New minor releases can indicate additional attributes / features in the model.

Author: I suspect we may want to treat these "well-known labels" the same way we treat extensions in CloudEvents. Meaning, anyone can propose new ones at any time and they might even change their definitions over time as people start to use them (or even remove them). So, it might be best to create a 2nd doc that has a different lifecycle than the spec itself and has a much lower bar for edits (and no official versioning). WDYT?

Reviewer 2: Yes, I like the idea! 

We also need to put the "compatiblewith" into the 3 CNCF provided models, so when you import them you already have them by default. That would make this label also more visible. 

I thought of another simplification: How about you can just add any URI, and each model defines which it is? You could also point to the model specification, which may be more user friendly e.g. https://github.com/xregistry/spec/blob/main/endpoint/spec.md

Reviewer 2: @duglin I can see three kinds of well-known labels: 1) labels on the meta-model (like `compatiblewith` and `modelversion` (?)), 2) labels on the instance level (not sure we have any) and 3) labels defined as part of a model, e.g. the endpoint model.

Here is a PR to add it to the extension specs: https://github.com/xregistry/spec/pull/225
